taurus/util/impl.py

```diff
@@ -25,19 +26,57 @@ def func(base_obj):
     return
 
 
+def _substitute(thing, sub, applies, visited: Dict[int, object] = None):
```

Comment:
Could you add a docstring for this method?
Also, why is the typehint for `visited` a Dict from int to object? Shouldn't it be `Dict[object, object]`?

Reply:
yeah that's from an old implementation, I will add a docstring and fix the type hint